src/utils/openCloseComponent.ts

```diff
@@ -8,20 +8,90 @@ export interface OpenCloseComponent {
    */
   readonly el: HTMLElement;
 
+  /**
+   * When true, the component opens.
+   */
+  open: boolean;
+
+  /**
+   * Specifies property on which active transition is watched for.
+   */
+  openTransitionProp: string;
+
+  /**
+   * Specifies element that the transition is allowed to emit on.
+   */
+  transitionEl: HTMLDivElement;
+
   /**
    * Defines method for `beforeOpen` event handler.
    */
   onBeforeOpen: () => void;
+
   /**
    * Defines method for `open` event handler:
    */
   onOpen: () => void;
+
   /**
    * Defines method for `beforeClose` event handler:
    */
   onBeforeClose: () => void;
+
   /**
    * Defines method for `close` event handler:
    */
   onClose: () => void;
 }
+
+const componentToTransitionListeners = new WeakMap<
+  OpenCloseComponent,
+  [typeof transitionStart, typeof transitionEnd]
+>();
+
+function transitionStart(event: TransitionEvent): void {
+  if (event.propertyName === this.openTransitionProp && event.target === this.transitionEl) {
+    this.open ? this.onBeforeOpen() : this.onBeforeClose();
+  }
+}
+
+function transitionEnd(event: TransitionEvent): void {
+  if (event.propertyName === this.openTransitionProp && event.target === this.transitionEl) {
+    this.open ? this.onOpen() : this.onClose();
+  }
+}
+/**
+ * Helper to keep track of transition listeners on setTransitionEl and connectedCallback on OpenCloseComponent components.
+ *
+ * @param component
+ */
+export function connectOpenCloseComponent(component: OpenCloseComponent): void {
+  if (component.transitionEl) {
+    const [start, end] = componentToTransitionListeners.get(component);
```

Comment:
Do you think that we should just call `disconnectOpenCloseComponent(component);` here and save some lines of code?